woocommerce/woocommerce-ios
# Filter Products screen

Filter list selector screen for #2037 

- [x] ⚠️ Update PR base to `develop` after https://github.com/woocommerce/woocommerce-ios/pull/2195 is merged ⚠️ 

## Changes

Apology for the size of the PR 🙇🏻‍♀️ the `BottomButtonContainerView` should be replaced by https://github.com/woocommerce/woocommerce-ios/pull/2210.

- Added `description` to `ProductType` so that the product type value can be displayed in the filter list selector screen
- Created `BottomButtonContainerView` that contains a primary CTA and is displayed at the bottom of a filter list screen
- Created `FilterListViewController` + xib as the container view controller that consists of 2 container views in a vertical stack view:
  - the top view contains a `UINavigationController` with a `ListSelectorViewController` with filter rows and their current values
    - Tapping on a filter row navigates to another `ListSelectorViewController` with the options, title, and selected value from the `FilterTypeViewModel`
  - the bottom view contains a CTA to filter the products with the given filters
<img width="443" alt="Screen Shot 2020-04-27 at 4 32 44 PM" src="https://user-images.githubusercontent.com/1945542/80351763-71e47380-88a5-11ea-8ee6-275b71a92569.png">

- Created `FilterTypeViewModel` that is used to render the row for the filter type and how the value list selector is configured
- Created `FilterListViewModel` protocol to configure `FilterListViewController` - it consists of an array of `FilterTypeViewModel` to configure the list of different filter types, handles all the actions (dismiss/filter/clear all/select a filter)
- Created `FilterProductListViewModel` to implement `FilterListViewModel` for Filter Products UI
- In `ProductsViewController`, created a `FilterProductListViewModel` and presented the filter list UI when the user taps on the "Filter" CTA in the toolbar

## Testing

- Go to the Products tab
- Tap on the "Filter" CTA in the toolbar below the WIP banner --> should see the Filter Products screen **without "Clear All" CTA in the navigation bar**
- Tap on "Stock status" row --> it should navigate to the stock status list selector screen
- Select a different stock status (e.g. "Out of stock") filter and tap "Show Products" at the bottom --> it should go back to the Products tab without any changes to the product list
- Tap on the "Filter" CTA again --> should see the Filter Products screen as in the screenshots below, **with "Clear All" CTA in the navigation bar**, the navigation bar title should say "Filters (1)", and the "Stock status" row should say the selected stock status
- Repeat the same steps for "Product status" and "Product type" rows and notice the navigation bar title should update to the number of active filters
- Tap "Clear all" CTA in the navigation bar --> all filters should be cleared to "Any"


## Example screenshots

Note for @Garance91540: the list will be updated to be plain style (no space above the cells) in a later task

mode \ screen | Filter Products - portrait | Filter Products - landscape
-- | -- | --
Dark | ![Simulator Screen Shot - iPhone 11 - 2020-04-28 at 11 52 41](https://user-images.githubusercontent.com/1945542/80445730-909b4680-8947-11ea-9c21-1011641dc08b.png) | ![Simulator Screen Shot - iPhone 11 - 2020-04-28 at 11 52 46](https://user-images.githubusercontent.com/1945542/80445734-92fda080-8947-11ea-8135-cccdf45fb87a.png)
Light | ![Simulator Screen Shot - iPhone 11 - 2020-04-28 at 11 52 57](https://user-images.githubusercontent.com/1945542/80445736-942ecd80-8947-11ea-9069-c0d82eadac6b.png) | ![Simulator Screen Shot - iPhone 11 - 2020-04-28 at 11 52 59](https://user-images.githubusercontent.com/1945542/80445787-b32d5f80-8947-11ea-9dc4-eb2207593207.png)


largest font size:

portrait | landscape
-- | --
![Simulator Screen Shot - iPhone 11 - 2020-04-28 at 11 53 43](https://user-images.githubusercontent.com/1945542/80445720-87aa7500-8947-11ea-9b37-4fe2cb710e40.png) | ![Simulator Screen Shot - iPhone 11 - 2020-04-28 at 11 53 50](https://user-images.githubusercontent.com/1945542/80445723-8aa56580-8947-11ea-9785-b2342e6237fa.png)

mode \ screen | filter products | filter products by stock status | filter products by product status | filter products by product type
-- | -- | -- | -- | --
Light | ![Simulator Screen Shot - iPhone 11 - 2020-04-28 at 18 35 39](https://user-images.githubusercontent.com/1945542/80477961-5e0d4000-897f-11ea-87d4-0b0912fb3179.png) | ![Simulator Screen Shot - iPhone 11 - 2020-04-28 at 18 35 26](https://user-images.githubusercontent.com/1945542/80477945-5a79b900-897f-11ea-96d9-a2af047fe2e4.png) | ![Simulator Screen Shot - iPhone 11 - 2020-04-28 at 18 31 07](https://user-images.githubusercontent.com/1945542/80477888-48981600-897f-11ea-9996-1406e00e451d.png) | ![Simulator Screen Shot - iPhone 11 - 2020-04-28 at 18 35 30](https://user-images.githubusercontent.com/1945542/80477958-5cdc1300-897f-11ea-909a-6122a31aaea9.png)
Dark | ![Simulator Screen Shot - iPhone 11 - 2020-04-28 at 18 35 49](https://user-images.githubusercontent.com/1945542/80477963-5ea5d680-897f-11ea-8fb5-ceb1e7f9bfa9.png) | ![Simulator Screen Shot - iPhone 11 - 2020-04-28 at 18 35 52](https://user-images.githubusercontent.com/1945542/80477966-5ea5d680-897f-11ea-8bc8-bddd495ba830.png) | ![Simulator Screen Shot - iPhone 11 - 2020-04-28 at 18 35 54](https://user-images.githubusercontent.com/1945542/80477968-5f3e6d00-897f-11ea-916f-eb4fa4804c37.png) | ![Simulator Screen Shot - iPhone 11 - 2020-04-28 at 18 35 56](https://user-images.githubusercontent.com/1945542/80477969-5fd70380-897f-11ea-8d41-586233023584.png)

iPad Air 2 (iOS 13.4)

mode \ screen | filter products | filter products by a filter
-- | -- | --
Dark (landscape) | ![Simulator Screen Shot - iPad Air 2 - 2020-04-29 at 16 04 19](https://user-images.githubusercontent.com/1945542/80573890-6e7ef280-8a33-11ea-9162-52dcd6f4471a.png) | ![Simulator Screen Shot - iPad Air 2 - 2020-04-29 at 16 04 03](https://user-images.githubusercontent.com/1945542/80573881-68891180-8a33-11ea-872f-f00eb391a019.png)
Dark (portrait) | ![Simulator Screen Shot - iPad Air 2 - 2020-04-29 at 16 04 24](https://user-images.githubusercontent.com/1945542/80573897-6fb01f80-8a33-11ea-8427-9d8fa81911c4.png) | ![Simulator Screen Shot - iPad Air 2 - 2020-04-29 at 16 04 27](https://user-images.githubusercontent.com/1945542/80573903-70e14c80-8a33-11ea-91fe-f245c338854f.png)
Light (landscape) | ![Simulator Screen Shot - iPad Air 2 - 2020-04-29 at 16 04 44](https://user-images.githubusercontent.com/1945542/80573917-73dc3d00-8a33-11ea-9b3d-1cede246add5.png) | ![Simulator Screen Shot - iPad Air 2 - 2020-04-29 at 16 04 47](https://user-images.githubusercontent.com/1945542/80573922-75a60080-8a33-11ea-9976-b9eb87935982.png)
Light (portrait) | ![Simulator Screen Shot - iPad Air 2 - 2020-04-29 at 16 04 42](https://user-images.githubusercontent.com/1945542/80573913-7343a680-8a33-11ea-930a-e6581b2de67a.png) | ![Simulator Screen Shot - iPad Air 2 - 2020-04-29 at 16 04 39](https://user-images.githubusercontent.com/1945542/80573910-72ab1000-8a33-11ea-8b7e-2eefa15d2a56.png)

Update release notes:
